### PR TITLE
dataclass_replace , curried reduce

### PR DESCRIPTION
**Pull request by molaxx**
added dataclass_replace to allow dataclass mutation in a functional way, added a sane version for a curried reduce

### CODE REVIEW

File: gamla/functional.py [hunk 1]
```diff
@@ -314,3 +315,32 @@ def update_in(d, keys, func, default=None, factory=dict):
...
+
+_R = TypeVar("_R")
+_E = TypeVar("_E")
+
+
+@toolz.curry
+def reduce(
+    reducer: Callable[[_R, _E], _R], initial_value: _R, elements: Iterable[_E],
+) -> _R:
+    return toolz.pipe(
```
Comment (nitzo) [inline gamla/functional.py:342]: not sure why we need the pipeline here ? isn't this as simple as `return functools.reduce(reducer, elements, initial_value)` ?

File: gamla/functional.py [hunk 2]
```diff
@@ -314,3 +315,32 @@ def update_in(d, keys, func, default=None, factory=dict):
     else:
         inner[k] = func(default)
     return rv
+
+
+@toolz.curry
+def dataclass_replace(
+    attr_name: Text, attr_mutator: Callable[[Any], Any], dataclass_instance
```
Comment (nitzo) [inline gamla/functional.py:322]: I don't know how you are using this but most of the times we just replace values (so the mutator would be `gamla.just(x)` ) maybe add a sugar `dataclass_replace_with_value` ? wdyt ?

File: gamla/functional.py (hunk 1 above)
Comment (uriva) [inline gamla/functional.py:342]: this seems like an overkill, why not just call `reduce` with the variables?

e.g. `return reduce(reducer, elements, initial_value)`

Comment (uriva) [reply gamla/functional.py:342]: lol @nitzo you beat me to it, sorry didn't see this is already reviewed

**Pull request merged**